### ISSUE TITLE
Metabox options are not saved in the Course post type

### ISSUE THREAD

**Issue by oceanwp** [gocodebox/lifterlms#359]
### 1. Reproduction Steps

Hello, I have a small problem with LifterLMS. I created a theme OceanWP, in the next official version of my theme, I would like it to be compatible with your plugin, so I modify some files of my theme and try to integrate my OceanWP Settings metabox which allows users to modify several theme options. The problem is that it does not work on the Course post type and I do not know at all why, it works very well on the Lesson post type.

To reproduce the issue:
1. Download my theme: https://github.com/oceanwp/oceanwp
2. Go to the Course post type of your plugin, modify any option in the OceanWP Settings metabox and save the item, you will see that no options are saved. In the Lesson post type, no problem at all.

Do you have any idea on why there is this issue?
Thank you very much.

### 2. Expected Behavior

That the OceanWP Settings options work as for the lesson post type.

### 3. Actual Behavior

The metabox option for the course post type reset to their default value when i save the item.

### 4. Error Messages / Logs

No error message

### 5. System Report

Wordpress
-------------------------------------------

Home Url: http://localhost/work
Site Url: http://localhost/work
Version: 4.8
Debug Mode: Yes
Locale: en_US
Multisite: No
Page For Posts: Blog (#31)
Page On Front: Home (#396)
Permalink Structure: /%postname%/
Show On Front: page
Wp Cron: Yes

Settings
-------------------------------------------

Version: 3.10.2
Db Version: 3.10.2
Course Catalog: Course Catalog (#1319)
Course Catalog Per Page: 9
Course Catalog Sorting: menu_order
Membership Catalog: Membership Catalog (#1320)
Membership Catalog Per Page: 9
Membership Catalog Sorting: menu_order
Site Membership: Not Set
Student Dashboard: My Courses (#1322)
Courses Endpoint: my-courses
Edit Endpoint: edit-account
Lost Password Endpoint: lost-password
Vouchers Endpoint: redeem-voucher
Autogenerate Username: yes
Password Strength Meter: yes
Minimum Password Strength: strong
Terms Required: no
Terms Page: Not Set
Checkout Names: required
Checkout Address: required
Checkout Phone: optional
Checkout Email Confirmation: yes
Open Registration: no
Registration Names: required
Registration Address: optional
Registration Phone: hidden
Registration Voucher: optional
Registration Email Confirmation: no
Account Names: required
Account Address: required
Account Phone: optional
Account Email Confirmation: yes
Checkout Page: Purchase (#1321)
Confirmation Endpoint: confirm-payment
Force Ssl Checkout: no
Country: US
Currency: USD
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Decimals: 2
Trim Zero Decimals: no
Recurring Payments: yes
Email From Address: oceanwptheme@gmail.com
Email From Name: OceanWP
Email Footer Text: 
Email Header Image: 
Cert Bg Width: 800
Cert Bg Height: 616
Cert Legacy Compat: no

Gateways
-------------------------------------------

Manual: Disabled
Manual Logging: 
Manual Order: 1

Server
-------------------------------------------

Mysql Version: 5.5.5
Php Curl: Yes
Php Default Timezone: UTC
Php Fsockopen: Yes
Php Max Input Vars: 1000
Php Max Upload Size: 64 MB
Php Memory Limit: 512M
Php Post Max Size: 64M
Php Soap: No
Php Suhosin: No
Php Time Limt: 300
Php Version: 7.0.15
Software: Apache/2.4.25 (Win32) OpenSSL/1.0.2j PHP/7.0.15
Wp Memory Limit: 40M

Theme
-------------------------------------------

Name: OceanWP
Version: 1.2.9
Themeuri: 
Authoruri: 
Template: oceanwp
Child Theme: No
Llms Support: No

Plugins
-------------------------------------------

Duplicate Post: 3.2
Elementor: 1.5.5
Elementor Pro: 1.5.6
LifterLMS: 3.10.2
NS Theme Check: 0.1.4
Ocean Demo Import: 1.0.4
Ocean Elementor Widgets: 1.0.6
Ocean Extra: 1.2.9
Regenerate Thumbnails: 2.2.6
Theme Check: 20160523.1
WooCommerce: 3.1.1
WordPress Database Reset: 3.0.2
WordPress Importer: 0.6.3

Integrations
-------------------------------------------

BbPress: No
BuddyPress: No

**Comment by thomasplevy** [comment]
@oceanwp stoked you're adding LifterLMS support -- i'll get you rolling to the best of my ability but I don't see any metaboxes after I install your theme

I browsed the source quickly and I can't find any mention of actual metabox classes that your lifterlms class is adding. Do I need to install a parent theme or some kind of plugin dependency to reproduce this?

**Comment by oceanwp** [comment]
Thank you for your prompt response :)
Sorry, I forgot to send you the latest version of the Ocean Extra plugin: https://www.dropbox.com/s/u98fy2eym8kl2hz/ocean-extra.zip?dl=0
So you have to use the theme on github (https://github.com/oceanwp/oceanwp) and the latest version of Ocean Extra that I just sent you.
You will see a OceanWP Settings metabox in Course and Lesson.
Thank you very much :)

**Comment by thomasplevy** [comment]
looks like the metabox framework is hooking into the creation of access plans --- access plans are created & updated when the course is

if you check your database you'll see the ocean meta fields are being stored on an access plan's postmeta record...

guessing that if you updated on a course with no access plans it'd probably work fine

I'm not going to make much more progress on this in the next few days due to a few big deadlines coming up.

I need to play around with this for a little while to determine if this is something that we are doing incorrectly -- i'm guessing it is but i really don't have the capacity to chug out an overhaul to fix this

I'll get back to you early next week with a better solution or proposed one

accept my apologies, I am sorry this is messed up (if it is our fault) and you've hit me at a bad time otherwise i'd fix it much faster.

**Comment by oceanwp** [comment]
I totally understand, I have no much time either :)
Take a look when you'll have time.
Thanks a lot :)

**Comment by thomasplevy** [comment]
@oceanwp I just wanted to let you know that I haven't forgotten about this issue I've just  been ignoring it (like that's much better, right?) I took a good hard look at it and think I have a solution that I'm going to put some time into over the weekend. Thanks for your continued patience!

**Comment by oceanwp** [comment]
Hi @thomasplevy, thanks a lot :)

**Comment by thomasplevy** [comment]
@oceanwp I cannot find a descent way to fix this from the LifterLMS side without what's proving to be a pretty major refactor. However, I've found a way to get it to work by adding a few lines of code to either your theme or plugin:

```php
add_action( 'llms_metabox_after_save_lifterlms-course-options', function( $post_id ) {
	butterbean()->update( $post_id );
} );
```

This will trigger your metabox save *after* LifterLMS course custom save functions run.

I ran a whole bunch of different tests against butterbean and it seems that there's a bit of wonkiness happening since it all runs of the `save_post` hook and that's called up to 7 times when a course with access plans is saved. It will save your ocean settings to the postmeta table for the first access plan. It's also doing some weirdness where there's no post_id set at all but still saving stuff.

Anyways, this isn't an IDEAL solution but it's something of a solution. It will result in extra metadata being stored on an access plan unnecessarily and it's a bit inefficient.

I tried to figure out a way to stop it from running against courses and then the above function could be used after courses (and access plans) are saved but I couldn't find a way to do it.

I'm going to refactor LifterLMS to save access plans via AJAX but I don't have time to go down that road right now.

Hopefully this gets you rolling despite being a bit wonky...
